iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.hpp

```diff
@@ -0,0 +1,199 @@
+///        This can be dynamic storage, static storage or anything else adhering to the allocation interface (cf.
+///        static_storage).
+
+/// @note This is not achievable with std::function and a custom allocator, as then the memory will still not
+///       be part of the object and copying (and moving may cause subtle issues). Hence a complete implementation
+///       is required.
+///       Furthermore the allocator support of std::function in the STL is deprecated.
+
+/// @tparam StorageType The type of internal storage to store the actual data.
+///                     Needs to provide allocate and deallocate functions.
+///                     See static_storage.hpp for a static memory version.
+/// @tparam ReturnType  The return type of the stored callable.
+/// @tparam Args        The arguments of the stored callable.
+template <typename StorageType, typename ReturnType, typename... Args>
+class storable_function<StorageType, signature<ReturnType, Args...>>
+{
+  public:
+    using signature_t = signature<ReturnType, Args...>;
+
+    storable_function() noexcept = default;
+
+    /// @brief construct from functor (including lambdas)
+    template <typename Functor,
+              typename = typename std::enable_if<std::is_class<Functor>::value
+                                                     && is_invocable_r<ReturnType, Functor, Args...>::value,
+                                                 void>::type>
+    storable_function(const Functor& functor) noexcept;
+
+    /// @brief construct from function pointer (including static functions)
+    storable_function(ReturnType (*function)(Args...)) noexcept;
+
+    /// @brief construct from object reference and member function
+    /// only a pointer to the object is stored for the call
+    template <typename T, typename = typename std::enable_if<std::is_class<T>::value, void>::type>
+    storable_function(T& object, ReturnType (T::*method)(Args...)) noexcept;
+
+    /// @brief construct from object reference and const member function
+    /// only a pointer to the object is stored for the call
+    template <typename T, typename = typename std::enable_if<std::is_class<T>::value, void>::type>
+    storable_function(const T& object, ReturnType (T::*method)(Args...) const) noexcept;
+
+    storable_function(const storable_function& other) noexcept;
+
+    storable_function(storable_function&& other) noexcept;
+
+    storable_function& operator=(const storable_function& rhs) noexcept;
+
+    storable_function& operator=(storable_function&& rhs) noexcept;
+
+    ~storable_function() noexcept;
+
+    /// @brief invoke the stored function
+    /// @param args arguments to invoke the stored function with
+    /// @return return value of the stored function
+    /// @note  Invoking the function if there is no stored function (i.e. operator bool returns false)
+    ///        leads to terminate being called.
```

Comment:
That is not how it is currently though, as far as I know. Therefore the comment should reflect this and will stay this way.

I prefer interface behavior description to be precise, may is not conductive to this (would be better to be able to state the conditions under which it will terminate).

It may work, or maybe not. Maybe it also blows up your computer...who knows. An exciting function to be sure, but may terminate is not a useful claim here.